BUG: to_file(filename, driver='kml') results in offset columns

- [x ] I have checked that this issue has not already been reported.

- [x ] I have confirmed this bug exists on the latest version of geopandas.

- [ ] (optional) I have confirmed this bug exists on the main branch of geopandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
import geopandas as gpd
from shapely.geometry import Point

# build a dict of columns: 2 text, 1 int, 1 float, and the geometry
data = {
    "text_a":   ["foo",   "bar",   "baz"],
    "text_b":   ["lorem", "ipsum", "dolor"],
    "int_val":  [1,       2,       3],
    "float_val":[0.1,     0.2,     0.3],
    "geometry": [
        Point(-105.0, 40.0),
        Point(-110.0, 45.0),
        Point(-115.0, 50.0),
    ],
}

# create the GeoDataFrame (with WGS84 lon/lat)
gdf = gpd.GeoDataFrame(data, crs="EPSG:4326")

gdf.to_file("blah.kml", driver="kml")
with open("blah.kml") as ff:
    print(ff.read())

```

#### Problem description

In the ExtendedData tag, the first element has the name of the first column but it has the value of the 3rd column.
```xml
  <Placemark id="blah.1">
	<name>foo</name>
	<description>lorem</description>
	<ExtendedData><SchemaData schemaUrl="#blah">
		<SimpleData name="text_a">1</SimpleData>
		<SimpleData name="text_b">0.1</SimpleData>
	</SchemaData></ExtendedData>
      <Point><coordinates>-105,40</coordinates></Point>
  </Placemark>
```

#### Expected Output
The column names should match their values and not exclude the last two columns.

```xml
  <Placemark id="blah.1">
	<name>foo</name>
	<description>lorem</description>
	<ExtendedData><SchemaData schemaUrl="#blah">
		<SimpleData name="text_a">foo</SimpleData>
		<SimpleData name="text_b">lorem</SimpleData>
        <SimpleData name="int_val">1</SimpleData>
        <SimpleData name="float_val">0.1</SimpleData>
	</SchemaData></ExtendedData>
      <Point><coordinates>-105,40</coordinates></Point>
  </Placemark>
```

Since the first two columns are represented in the name and description tag, I'm OK with excluding them from ExtendedData but the important thing is that it isn't naming the first SimpleData by the first column but putting the third column's values there.

Using Fiona works though.

If I do
```python
import fiona
fiona.drvsupport.supported_drivers['KML'] = 'rw'
gdf.to_file("blah.kml", driver="KML", engine="fiona")
with open("blah.kml") as ff:
    print(ff.read())
```

```xml
	<ExtendedData><SchemaData schemaUrl="#blah">
		<SimpleData name="text_a">foo</SimpleData>
		<SimpleData name="text_b">lorem</SimpleData>
		<SimpleData name="int_val">1</SimpleData>
		<SimpleData name="float_val">0.1</SimpleData>
	</SchemaData></ExtendedData>
```
fiona doesn't populate a name or description tag which is fine with me.

#### Output of ``geopandas.show_versions()``

<details>



SYSTEM INFO
-----------
python     : 3.12.2 (main, Feb 25 2024, 16:35:05) [GCC 11.4.0]
executable : /home/dean/pyjs/generic/.venv/bin/python
machine    : Linux-6.6.87.2-microsoft-standard-WSL2-x86_64-with-glibc2.35

GEOS, GDAL, PROJ INFO
---------------------
GEOS       : 3.11.4
GEOS lib   : None
GDAL       : 3.10.3
GDAL data dir: /home/dean/pyjs/generic/.venv/lib/python3.12/site-packages/pyogrio/gdal_data/
PROJ       : 9.3.0
PROJ data dir: /home/dean/pyjs/generic/.venv/lib/python3.12/site-packages/pyproj/proj_dir/share/proj

PYTHON DEPENDENCIES
-------------------
geopandas  : 1.1.1
numpy      : 2.2.3
pandas     : 2.2.2
pyproj     : 3.6.1
shapely    : 2.0.7
pyogrio    : 0.11.0
geoalchemy2: 0.15.2
geopy      : None
matplotlib : 3.9.2
mapclassify: None
fiona      : 1.10.1
psycopg    : 3.2.3
psycopg2   : 2.9.10 (dt dec pq3 ext lo64)
pyarrow    : 20.0.0

</details>

This was not an easy fix but I think I found a fix for this issue. I wrote a test to check my solution and that is passing on the branch with the fix and failing on main.  

```python
@PYOGRIO_KML_MARK
def test_to_file_kml_pyogrio_keeps_extended_data_aligned(tmp_path):
    df = GeoDataFrame(
        {
            "text_a": ["foo", "bar"],
            "text_b": ["lorem", "ipsum"],
            "int_val": [1, 2],
            "float_val": [0.1, 0.2],
            "geometry": [Point(-105, 40), Point(-110, 45)],
        },
        crs="EPSG:4326",
    )
    filename = tmp_path / "test.kml"

    df.to_file(filename, driver="kml", engine="pyogrio")

    import xml.etree.ElementTree as ET

    root = ET.parse(filename).getroot()
    ns = {"kml": "http://www.opengis.net/kml/2.2"}
    simple_fields = root.findall(".//kml:SimpleField", ns)
    simple_data = root.findall(".//kml:Placemark[1]//kml:SimpleData", ns)

    assert [field.get("name") for field in simple_fields] == [
        "text_a",
        "text_b",
        "int_val",
        "float_val",
    ]
    assert [(data.get("name"), data.text) for data in simple_data] == [
        ("text_a", "foo"),
        ("text_b", "lorem"),
        ("int_val", "1"),
        ("float_val", "0.1"),
    ]
```

But there are two tests that fails though, both on main and on my branch with the fix. So it is unlikely they are related if you ask me.

This is the test result with the two failed tests:
```bash
============================= test session starts ==============================
platform linux -- Python 3.12.13, pytest-9.0.3, pluggy-1.6.0
rootdir: /home/cryptofarian/python/geopandas
configfile: pyproject.toml
plugins: xdist-3.8.0, doctestplus-1.7.1, cov-7.1.0
collected 2916 items

geopandas/io/tests/test_arrow.py ....................................... [  1%]
........................................s............................... [  3%]
..............................s                                          [  4%]
geopandas/io/tests/test_file.py ........................................ [  6%]
..ss....ssss....ss....ssss.............................................. [  8%]
........................................................x...x..x...x.... [ 11%]
....xx........................................................s......... [ 13%]
...........................s.........ss...                               [ 15%]
geopandas/io/tests/test_file_geom_types_drivers.py ..................... [ 15%]
........................................................................ [ 18%]
...................                                                      [ 18%]
geopandas/io/tests/test_geoarrow.py .................................... [ 20%]
..........................................................               [ 22%]
geopandas/io/tests/test_infer_schema.py .......................          [ 22%]
geopandas/io/tests/test_pickle.py sss.                                   [ 23%]
geopandas/io/tests/test_sql.py sssssssssssssssssssssssssssssssssssssssss [ 24%]
s                                                                        [ 24%]
geopandas/tests/test_api.py .                                            [ 24%]
geopandas/tests/test_array.py .......................................... [ 25%]
..........................................................s.....ss...... [ 28%]
                                                                         [ 28%]
geopandas/tests/test_compat.py .........                                 [ 28%]
geopandas/tests/test_config.py ...                                       [ 28%]
geopandas/tests/test_crs.py ............................................ [ 30%]
..............................                                           [ 31%]
geopandas/tests/test_datasets.py ....                                    [ 31%]
geopandas/tests/test_decorator.py ....                                   [ 31%]
geopandas/tests/test_dissolve.py ....................                    [ 32%]
geopandas/tests/test_explore.py ........................................ [ 33%]
.                                                                        [ 33%]
geopandas/tests/test_extension_array.py ................................ [ 34%]
........................................................................ [ 37%]
.xx..................x.................................................. [ 39%]
.................................xx........................x............ [ 42%]
........sssssssssssssssssssss........................................... [ 44%]
.................................ss..................................... [ 47%]
................ss...................................................... [ 49%]
.....ssssssssss..............................ss                          [ 51%]
geopandas/tests/test_geocode.py ........                                 [ 51%]
geopandas/tests/test_geodataframe.py ................................... [ 52%]
........................................................................ [ 55%]
....................                                                     [ 55%]
geopandas/tests/test_geom_methods.py ................................... [ 57%]
........................................................................ [ 59%]
........................................................................ [ 62%]
....................                                                     [ 62%]
geopandas/tests/test_geoseries.py ...................................... [ 64%]
.............s...........................                                [ 65%]
geopandas/tests/test_merge.py ........                                   [ 65%]
geopandas/tests/test_op_output_types.py ......ss........................ [ 66%]
....                                                                     [ 66%]
geopandas/tests/test_overlay.py ..........................s............. [ 68%]
s....................................................................... [ 70%]
..................                                                       [ 71%]
geopandas/tests/test_pandas_accessor.py .....                            [ 71%]
geopandas/tests/test_pandas_methods.py ................................. [ 72%]
........s............                                                    [ 73%]
geopandas/tests/test_plotting.py ....................................... [ 74%]
......F.....F........................................................... [ 77%]
.........................................................                [ 79%]
geopandas/tests/test_show_versions.py ....                               [ 79%]
geopandas/tests/test_sindex.py ................s........................ [ 80%]
........................................................................ [ 83%]
..................s.......x............................................. [ 85%]
...................................................x.................... [ 88%]
.......................                                                  [ 88%]
geopandas/tests/test_testing.py ........                                 [ 89%]
geopandas/tests/test_types.py ...............                            [ 89%]
geopandas/tools/tests/test_clip.py ..................................... [ 91%]
....................................................                     [ 92%]
geopandas/tools/tests/test_hilbert_curve.py ......                       [ 93%]
geopandas/tools/tests/test_random.py .................                   [ 93%]
geopandas/tools/tests/test_sjoin.py .................................... [ 94%]
........................................................................ [ 97%]
..............xs.......................................................  [ 99%]
geopandas/tools/tests/test_tools.py ........                             [100%]

=================================== FAILURES ===================================
____________ TestPolygonPlotting.test_plot_polygon_with_holes[png] _____________

args = (<geopandas.tests.test_plotting.TestPolygonPlotting object at 0x7f3694a950d0>,)
kwds = {'extension': 'png', 'request': <FixtureRequest for <Function test_plot_polygon_with_holes[png]>>}

    @wraps(func)
    def inner(*args, **kwds):
        with self._recreate_cm():
>           return func(*args, **kwds)
                   ^^^^^^^^^^^^^^^^^^^

/usr/lib/python3.12/contextlib.py:81: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
venv/lib/python3.12/site-packages/matplotlib/testing/compare.py:475: in compare_images
    rms = calculate_rms(expected_image, actual_image)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

expected_image = array([[[255, 255, 255],
        [255, 255, 255],
        [255, 255, 255],
        ...,
        [255, 255, 255],
     ...  ...,
        [255, 255, 255],
        [255, 255, 255],
        [255, 255, 255]]], shape=(1577, 1577, 3), dtype=int16)
actual_image = array([[[255, 255, 255],
        [255, 255, 255],
        [255, 255, 255],
        ...,
        [255, 255, 255],
     ...  ...,
        [255, 255, 255],
        [255, 255, 255],
        [255, 255, 255]]], shape=(1578, 1578, 3), dtype=int16)

    def calculate_rms(expected_image, actual_image):
        """
        Calculate the per-pixel errors, then compute the root mean square error.
        """
        if expected_image.shape != actual_image.shape:
>           raise ImageComparisonFailure(
                f"Image sizes do not match expected size: {expected_image.shape} "
                f"actual size {actual_image.shape}")
E           matplotlib.testing.exceptions.ImageComparisonFailure: Image sizes do not match expected size: (1577, 1577, 3) actual size (1578, 1578, 3)

venv/lib/python3.12/site-packages/matplotlib/testing/compare.py:372: ImageComparisonFailure
__________ TestPolygonPlotting.test_multipolygons_with_interior[png] ___________

args = (<geopandas.tests.test_plotting.TestPolygonPlotting object at 0x7f3694a82660>,)
kwds = {'extension': 'png', 'request': <FixtureRequest for <Function test_multipolygons_with_interior[png]>>}

    @wraps(func)
    def inner(*args, **kwds):
        with self._recreate_cm():
>           return func(*args, **kwds)
                   ^^^^^^^^^^^^^^^^^^^

/usr/lib/python3.12/contextlib.py:81: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
venv/lib/python3.12/site-packages/matplotlib/testing/compare.py:475: in compare_images
    rms = calculate_rms(expected_image, actual_image)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

expected_image = array([[[255, 255, 255],
        [255, 255, 255],
        [255, 255, 255],
        ...,
        [255, 255, 255],
     ...  ...,
        [255, 255, 255],
        [255, 255, 255],
        [255, 255, 255]]], shape=(1577, 1577, 3), dtype=int16)
actual_image = array([[[255, 255, 255],
        [255, 255, 255],
        [255, 255, 255],
        ...,
        [255, 255, 255],
     ...  ...,
        [255, 255, 255],
        [255, 255, 255],
        [255, 255, 255]]], shape=(1578, 1578, 3), dtype=int16)

    def calculate_rms(expected_image, actual_image):
        """
        Calculate the per-pixel errors, then compute the root mean square error.
        """
        if expected_image.shape != actual_image.shape:
>           raise ImageComparisonFailure(
                f"Image sizes do not match expected size: {expected_image.shape} "
                f"actual size {actual_image.shape}")
E           matplotlib.testing.exceptions.ImageComparisonFailure: Image sizes do not match expected size: (1577, 1577, 3) actual size (1578, 1578, 3)

venv/lib/python3.12/site-packages/matplotlib/testing/compare.py:372: ImageComparisonFailure
=============================== warnings summary ===============================
geopandas/tests/test_geom_methods.py: 204 warnings
  /home/cryptofarian/python/geopandas/venv/lib/python3.12/site-packages/pointpats/random.py:395: DeprecationWarning: `row_stack` alias is deprecated. Use `np.vstack` directly.
    clusters[i_cluster] = numpy.row_stack((seed, candidates))

geopandas/tests/test_geom_methods.py: 102 warnings
  /home/cryptofarian/python/geopandas/venv/lib/python3.12/site-packages/pointpats/random.py:396: DeprecationWarning: `row_stack` alias is deprecated. Use `np.vstack` directly.
    result[i_replication] = numpy.row_stack(clusters)

geopandas/tests/test_geom_methods.py: 204 warnings
  /home/cryptofarian/python/geopandas/venv/lib/python3.12/site-packages/pointpats/random.py:480: DeprecationWarning: `row_stack` alias is deprecated. Use `np.vstack` directly.
    clusters[i_cluster] = numpy.row_stack((center, candidates))

geopandas/tests/test_geom_methods.py: 102 warnings
  /home/cryptofarian/python/geopandas/venv/lib/python3.12/site-packages/pointpats/random.py:481: DeprecationWarning: `row_stack` alias is deprecated. Use `np.vstack` directly.
    result[i_replication] = numpy.row_stack(clusters)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED geopandas/tests/test_plotting.py::TestPolygonPlotting::test_plot_polygon_with_holes[png]
FAILED geopandas/tests/test_plotting.py::TestPolygonPlotting::test_multipolygons_with_interior[png]
= 2 failed, 2787 passed, 112 skipped, 15 xfailed, 612 warnings in 99.73s (0:01:39) =
```

Can I post a PR with this fix if this looks okay for you guys?

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -6,6 +6,7 @@
 
 -   Fix writing empty string category columns with arrow fails (#621).
 -   Fix overwriting a corrupt fileGDB directory (#600).
+-   Fix attribute data being incorrectly written with KML driver (#650).
 
 ## 0.12.1 (2025-11-28)
 

--- a/pyogrio/_io.pyx
+++ b/pyogrio/_io.pyx
@@ -2920,14 +2920,18 @@ def ogr_write(
         # Get the field indexes for the fields we need to write to, as they might be
         # different than the order they were created, e.g. when using the KML driver.
         ogr_featuredef = OGR_L_GetLayerDefn(ogr_layer)
+
         field_indexes = []
-        for i in range(num_fields):
-            index = OGR_FD_GetFieldIndex(ogr_featuredef, fields[i].encode(encoding))
-            if index < 0:
-                raise FieldError(
-                    f"Could not find field index for field '{fields[i]}'"
-                )
-            field_indexes.append(index)
+        if OGR_FD_GetFieldCount(ogr_featuredef) == num_fields:
+            field_indexes = list(range(num_fields))
+        else:
+            for i in range(num_fields):
+                index = OGR_FD_GetFieldIndex(ogr_featuredef, fields[i].encode(encoding))
+                if index < 0:
+                    raise FieldError(
+                        f"Could not find field index for field '{fields[i]}'"
+                    )
+                field_indexes.append(index)
 
         # Create the features
         supports_transactions = OGR_L_TestCapability(ogr_layer, OLCTransactions)

--- a/pyogrio/_io.pyx
+++ b/pyogrio/_io.pyx
@@ -2917,9 +2917,19 @@ def ogr_write(
                         OGR_Fld_Destroy(ogr_fielddef)
                         ogr_fielddef = NULL
 
-        # Create the features
+        # Get the field indexes for the fields we need to write to, as they might be
+        # different than the order they were created, e.g. when using the KML driver.
         ogr_featuredef = OGR_L_GetLayerDefn(ogr_layer)
+        field_indexes = []
+        for i in range(num_fields):
+            index = OGR_FD_GetFieldIndex(ogr_featuredef, fields[i].encode(encoding))
+            if index < 0:
+                raise FieldError(
+                    f"Could not find field index for field '{fields[i]}'"
+                )
+            field_indexes.append(index)
 
+        # Create the features
         supports_transactions = OGR_L_TestCapability(ogr_layer, OLCTransactions)
         if supports_transactions:
             start_transaction(ogr_dataset, 0)
@@ -2987,17 +2997,18 @@ def ogr_write(
             for field_idx in range(num_fields):
                 field_value = field_data[field_idx][i]
                 field_type = field_types[field_idx][0]
+                field_index = field_indexes[field_idx]
 
                 mask = field_mask[field_idx]
                 if mask is not None and mask[i]:
-                    OGR_F_SetFieldNull(ogr_feature, field_idx)
+                    OGR_F_SetFieldNull(ogr_feature, field_index)
 
                 elif field_type == OFTString:
                     if (
                         field_value is None
                         or (isinstance(field_value, float) and isnan(field_value))
                     ):
-                        OGR_F_SetFieldNull(ogr_feature, field_idx)
+                        OGR_F_SetFieldNull(ogr_feature, field_index)
 
                     else:
                         if not isinstance(field_value, str):
@@ -3005,7 +3016,7 @@ def ogr_write(
 
                         try:
                             value_b = field_value.encode(encoding)
-                            OGR_F_SetFieldString(ogr_feature, field_idx, value_b)
+                            OGR_F_SetFieldString(ogr_feature, field_index, value_b)
 
                         except AttributeError:
                             raise ValueError(
@@ -3017,25 +3028,25 @@ def ogr_write(
                             raise
 
                 elif field_type == OFTInteger:
-                    OGR_F_SetFieldInteger(ogr_feature, field_idx, field_value)
+                    OGR_F_SetFieldInteger(ogr_feature, field_index, field_value)
 
                 elif field_type == OFTInteger64:
-                    OGR_F_SetFieldInteger64(ogr_feature, field_idx, field_value)
+                    OGR_F_SetFieldInteger64(ogr_feature, field_index, field_value)
 
                 elif field_type == OFTReal:
                     if nan_as_null and isnan(field_value):
-                        OGR_F_SetFieldNull(ogr_feature, field_idx)
+                        OGR_F_SetFieldNull(ogr_feature, field_index)
                     else:
-                        OGR_F_SetFieldDouble(ogr_feature, field_idx, field_value)
+                        OGR_F_SetFieldDouble(ogr_feature, field_index, field_value)
 
                 elif field_type == OFTDate:
                     if np.isnat(field_value):
-                        OGR_F_SetFieldNull(ogr_feature, field_idx)
+                        OGR_F_SetFieldNull(ogr_feature, field_index)
                     else:
                         datetime = field_value.item()
                         OGR_F_SetFieldDateTimeEx(
                             ogr_feature,
-                            field_idx,
+                            field_index,
                             datetime.year,
                             datetime.month,
                             datetime.day,
@@ -3047,7 +3058,7 @@ def ogr_write(
 
                 elif field_type == OFTDateTime:
                     if np.isnat(field_value):
-                        OGR_F_SetFieldNull(ogr_feature, field_idx)
+                        OGR_F_SetFieldNull(ogr_feature, field_index)
                     else:
                         datetime = field_value.astype("datetime64[ms]").item()
                         tz_array = gdal_tz_offsets.get(fields[field_idx], None)
@@ -3057,7 +3068,7 @@ def ogr_write(
                             gdal_tz = tz_array[i]
                         OGR_F_SetFieldDateTimeEx(
                             ogr_feature,
-                            field_idx,
+                            field_index,
                             datetime.year,
                             datetime.month,
                             datetime.day,

--- a/pyogrio/_ogr.pxd
+++ b/pyogrio/_ogr.pxd
@@ -327,6 +327,8 @@ cdef extern from "ogr_api.h":
     OGRFeatureDefnH     OGR_FD_Create(const char *name)
     int                 OGR_FD_GetFieldCount(OGRFeatureDefnH featuredefn)
     OGRFeatureDefnH     OGR_FD_GetFieldDefn(OGRFeatureDefnH featuredefn, int n)
+    int                 OGR_FD_GetFieldIndex(OGRFeatureDefnH featuredefn,
+                                             const char *name)
     OGRwkbGeometryType  OGR_FD_GetGeomType(OGRFeatureDefnH featuredefn)
 
     OGRFieldDefnH   OGR_Fld_Create(const char *name, OGRFieldType fieldtype)

--- a/pyogrio/tests/test_geopandas_io.py
+++ b/pyogrio/tests/test_geopandas_io.py
@@ -3059,6 +3059,9 @@ def test_write_kml(tmp_path, kml_driver, use_arrow):
 
     Test added when fixing https://github.com/geopandas/geopandas/issues/3609
     """
+    if kml_driver not in list_drivers():
+        pytest.skip(f"{kml_driver} driver not available")
+
     df = gp.GeoDataFrame(
         {"col_1": [1.0, 2.0], "col_2": [3.0, 4.0], "col_3": [5.0, 6.0]},
         geometry=[Point(0, 0), Point(1, 1)],

--- a/pyogrio/tests/test_geopandas_io.py
+++ b/pyogrio/tests/test_geopandas_io.py
@@ -3080,7 +3080,11 @@ def test_write_kml(tmp_path, kml_driver, use_arrow):
     try:
         result_df = result_df[df.columns]
     except KeyError as ex:
-        raise RuntimeError(f"Keyerror for {df.columns=} on {result_df=}") from ex
+        with open(output_path) as f:
+            content = f.read()
+        raise RuntimeError(
+            f"Keyerror for {df.columns=} on {result_df=}, {content=}"
+        ) from ex
     assert_geodataframe_equal(result_df, df, check_index_type=False)
 
 

--- a/pyogrio/tests/test_geopandas_io.py
+++ b/pyogrio/tests/test_geopandas_io.py
@@ -1198,7 +1198,8 @@ def test_write_read_column_names_casing(tmp_path, ext, use_arrow, expected_resul
 
     filename = tmp_path / f"test_duplicate_columns{ext}"
     if expected_result == "error":
-        with pytest.raises(Exception, match="Error .* field .*'COL' .*"):
+        with pytest.raises(Exception, match="COL"):
+            # The message depends on the driver and GDAL version, keep match simple.
             write_dataframe(gdf, filename, use_arrow=use_arrow)
         return
 

--- a/pyogrio/tests/test_geopandas_io.py
+++ b/pyogrio/tests/test_geopandas_io.py
@@ -3045,6 +3045,39 @@ def test_arrow_enable_with_environment_variable(tmp_path):
             _ = read_dataframe(test_path, encoding="cp1252")
 
 
+@pytest.mark.requires_arrow_write_api
+@pytest.mark.parametrize("kml_driver", ["LIBKML", "KML"])
+def test_write_kml(tmp_path, kml_driver, use_arrow):
+    """Test writing a KML file.
+
+    A KML file is a bit of a special case, because when it is written, some extra
+    columns are added automatically in the layer definition: "Name" and "Description".
+    Because these extra columns are the first columns in the layer definition, it is
+    important to explicitly check the column index when writing values to fields as
+    you cannot rely on the index to be the same as the order fields were added to a
+    layer.
+
+    Test added when fixing https://github.com/geopandas/geopandas/issues/3609
+    """
+    df = gp.GeoDataFrame(
+        {"col_1": [1.0, 2.0], "col_2": [3.0, 4.0], "col_3": [5.0, 6.0]},
+        geometry=[Point(0, 0), Point(1, 1)],
+        crs="EPSG:4326",
+    )
+
+    output_path = tmp_path / "test.kml"
+    write_dataframe(df, output_path, driver=kml_driver, use_arrow=use_arrow)
+
+    assert output_path.exists()
+
+    result_df = read_dataframe(output_path)
+
+    # In a KML, there are several columns that are added automagically... so only check
+    # the columns we wrote.
+    result_df = result_df[df.columns]
+    assert_geodataframe_equal(result_df, df, check_index_type=False)
+
+
 @pytest.mark.filterwarnings("ignore:File /vsimem:RuntimeWarning")
 @pytest.mark.parametrize("driver", ["GeoJSON", "GPKG"])
 def test_write_memory(naturalearth_lowres, driver):

--- a/pyogrio/tests/test_geopandas_io.py
+++ b/pyogrio/tests/test_geopandas_io.py
@@ -292,12 +292,13 @@ def test_read_geojson_error(naturalearth_lowres_geojson, use_arrow):
 
 @pytest.mark.skipif(
     "LIBKML" not in list_drivers(),
-    reason="LIBKML driver is not available and is needed to read simpledata element",
+    reason="LIBKML driver is not available and is needed to read attribute columns",
 )
 def test_read_kml_simpledata(kml_file, use_arrow):
-    """Test reading a KML file with a simpledata element.
+    """Test reading a KML file with an attribute column.
 
-    Simpledata elements are only read by the LibKML driver, not the KML driver.
+    Attribute columns (="Simpledata" elements in the .kml) are only read by the LibKML
+    driver, not the KML driver.
     """
     gdf = read_dataframe(kml_file, use_arrow=use_arrow)
 
@@ -3103,7 +3104,7 @@ def test_arrow_enable_with_environment_variable(tmp_path):
 @pytest.mark.parametrize("kml_driver", ["LIBKML", "KML"])
 @pytest.mark.skipif(
     "LIBKML" not in list_drivers(),
-    reason="LIBKML driver is not available and is needed to read simpledata element",
+    reason="LIBKML driver is not available and is needed to read attribute columns",
 )
 def test_write_kml(tmp_path, kml_driver, use_arrow):
     """Test writing a KML file.

--- a/pyogrio/tests/test_geopandas_io.py
+++ b/pyogrio/tests/test_geopandas_io.py
@@ -3047,6 +3047,10 @@ def test_arrow_enable_with_environment_variable(tmp_path):
 
 @pytest.mark.requires_arrow_write_api
 @pytest.mark.parametrize("kml_driver", ["LIBKML", "KML"])
+@pytest.mark.skipif(
+    "LIBKML" not in list_drivers(),
+    reason="LIBKML driver is not available and is needed to read simpledata element",
+)
 def test_write_kml(tmp_path, kml_driver, use_arrow):
     """Test writing a KML file.
 
@@ -3077,14 +3081,7 @@ def test_write_kml(tmp_path, kml_driver, use_arrow):
 
     # In a KML, there are several columns that are added automagically... so only check
     # the columns we wrote.
-    try:
-        result_df = result_df[df.columns]
-    except KeyError as ex:
-        with open(output_path) as f:
-            content = f.read()
-        raise RuntimeError(
-            f"Keyerror for {df.columns=} on {result_df=}, {content=}"
-        ) from ex
+    result_df = result_df[df.columns]
     assert_geodataframe_equal(result_df, df, check_index_type=False)
 
 

--- a/pyogrio/tests/test_geopandas_io.py
+++ b/pyogrio/tests/test_geopandas_io.py
@@ -1169,6 +1169,59 @@ def test_write_read_datetime_utc(
         assert_geodataframe_equal(result, df)
 
 
+@pytest.mark.requires_arrow_write_api
+@pytest.mark.parametrize(
+    "ext, use_arrow, expected_result",
+    [
+        (".gpkg", False, "error"),
+        (".geojson", False, "supported"),
+        (".geojsonl", False, "supported"),
+        (".shp", False, "second_column_dropped"),
+        (".gpkg", True, "error"),
+        (".geojson", True, "second_column_overwrites_first"),
+        (".geojsonl", True, "second_column_overwrites_first"),
+        (".shp", True, "second_column_dropped"),
+    ],
+)
+def test_write_read_column_names_casing(tmp_path, ext, use_arrow, expected_result):
+    """Test writing and reading a file with column names that only differ in casing.
+
+    Probably never a good idea to use multiple columns with the same name but different
+    casing, but at least this test documents the current behaviour.
+
+    With arrow, this never seems to be supported.
+    """
+    df = pd.DataFrame(
+        {"col": [1, 2], "COL": [3, 4], "geometry": [Point(0, 0), Point(1, 1)]}
+    )
+    gdf = gp.GeoDataFrame(df, geometry="geometry", crs="EPSG:4326")
+
+    filename = tmp_path / f"test_duplicate_columns{ext}"
+    if expected_result == "error":
+        with pytest.raises(Exception, match="Error .* field .*'COL' .*"):
+            write_dataframe(gdf, filename, use_arrow=use_arrow)
+        return
+
+    write_dataframe(gdf, filename, use_arrow=use_arrow)
+
+    result = read_dataframe(filename, use_arrow=use_arrow)
+    assert "geometry" in result.columns
+    assert "col" in result.columns
+
+    if expected_result == "supported":
+        assert "COL" in result.columns
+        assert_series_equal(result["col"], df["col"], check_dtype=False)
+        assert_series_equal(result["COL"], df["COL"], check_dtype=False)
+    elif expected_result == "second_column_dropped":
+        assert_series_equal(result["col"], df["col"], check_dtype=False)
+        assert "COL" not in result.columns
+    elif expected_result == "second_column_overwrites_first":
+        assert_series_equal(result["col"], df["COL"].rename("col"), check_dtype=False)
+        assert "COL" not in result.columns
+    else:
+        raise AssertionError(f"Unhandled value for {expected_result=}")
+
+
 def test_read_null_values(tmp_path, use_arrow):
     filename = tmp_path / "test_null_values_no_geometry.gpkg"
 

--- a/pyogrio/tests/test_geopandas_io.py
+++ b/pyogrio/tests/test_geopandas_io.py
@@ -3077,7 +3077,10 @@ def test_write_kml(tmp_path, kml_driver, use_arrow):
 
     # In a KML, there are several columns that are added automagically... so only check
     # the columns we wrote.
-    result_df = result_df[df.columns]
+    try:
+        result_df = result_df[df.columns]
+    except KeyError as ex:
+        raise RuntimeError(f"Keyerror for {df.columns=} on {result_df=}") from ex
     assert_geodataframe_equal(result_df, df, check_index_type=False)
 
 

